client/client.go

```
@@ -803,53 +780,148 @@ func (c *Client) Call(ctx context.Context, procedure string, options wamp.Dict,
+			close(progDone)
+		}()
+	}
+
+	if finalrescb == nil {
+		return nil, errors.New("no result handler provided")
+	}
+
+	id := c.idGen.Next()
+	c.expectReply(id)
+	message := &wamp.Call{
+		Request:   id,
+		Procedure: wamp.URI(procedure),
+		Options:   options,
+	}
+
+	err := c.prepareCallPayloadMessage(message, options, args, kwargs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.sess.Send(message)
+
+	// As we can not wait for final result and must return control to the client code
+	// we wait for result in goroutine
+	go func() {
+		// Wait to receive RESULT message.
+		msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)
+
+		// Finish handling any remaining progressive results before returning the
+		// final result.
+		if progcb != nil {
+			close(progChan)
+			<-progDone
+		}
+
+		if err != nil {
+			finalrescb(nil, err)
+			return
+		}
+
+		switch msg := msg.(type) {
+		case *wamp.Result:
+
+			err := c.prepareCallResultMessage(msg)
 
 			if err != nil {
-				return nil, err
+				finalrescb(nil, err)
+				return
 			}
 
-			msg.Arguments = args
-			msg.ArgumentsKw = kwargs
+			finalrescb(msg, nil)
+			return
+		case *wamp.Error:
+			finalrescb(nil, RPCError{msg, procedure})
+			return
+		default:
+			finalrescb(nil, unexpectedMsgError(msg, wamp.RESULT))
+			return
+		}
+	}()
+
+	// now we need to return callback for feeding the progressive data
+	sendProgressiveData := func(options wamp.Dict, args wamp.List, kwargs wamp.Dict, isLast bool) error {
+		if options == nil {
+			options = wamp.Dict{}
 		}
-		return msg, nil
-	case *wamp.Error:
-		return nil, RPCError{msg, procedure}
-	default:
-		return nil, unexpectedMsgError(msg, wamp.RESULT)
+
+		options[wamp.OptProgress] = isLast
+
+		message := &wamp.Call{
+			Request:   id,
+			Procedure: wamp.URI(procedure),
+			Options:   options,
+		}
+
+		err := c.prepareCallPayloadMessage(message, options, args, kwargs)
+
+		if err != nil {
+			return err
+		}
+
+		c.sess.Send(message)
+
+		return nil
```

I think you should just return `c.sess.Send(message)`, assuming return signatures match.

---

Well, it may return an error, which is never checked for across the whole project. So I'll leave it untouched for now

---

I mean, it's weird. Either:

1) We should change `Send` not to return an `error` value.
2) Or we should return the error value it returns.
3) Or we should use something like `_ = c.sess.Send(message)` to explicitly *ignore* the error value.

What we have now looks like an omission, not deliberate design.

---

Yes, you are right of course!
Well, i'll think, maybe adding explicit ignore — is ok. As it is just sending message through the wire using underlying connection.